jfix-stdlib-batching/src/main/java/ru/fix/stdlib/batching/BatchingManagerThread.java

```diff
@@ -104,110 +108,171 @@ public void start() {
-                        try {
-                            if (waitTime <= 0) {
-                                waitTime = 1;
-                            }
-                            waitForOperationLock.wait(waitTime);
-                            log.trace("leave wait section");
-                        } catch (InterruptedException exc) {
-                            log.trace(THREAD_INTERRUPTED_MESSAGE, exc);
-                            Thread.currentThread().interrupt();
-                            return;
-                        }
-                    }
+                    threadInterrupted = metricsProvider.profileBatchManagerThreadAwaitOperation(
+                            this::calculateTimeForSleepAndWaitForOperations
+                    );
+
+                    if (threadInterrupted) return;
                 }
             }
 
             if (!isShutdown.get() && buffer != null) {
                 log.trace("Start batch processor");
-                batchProcessorPool.execute(new BatchProcessor<>(config,
+                executeBatchForBuffer(buffer, key);
+            }
+        }
+
+        metricsProvider.detachOperationQueueSizeIndicators();
+        log.trace("BatchProcessorManager thread stopped.");
+    }
+
+    private void executeBatchForBuffer(List<Operation<PayloadT>> buffer, KeyT key) {
+        batchProcessorPool.execute(
+                new BatchProcessor<>(
+                        config,
                         buffer,
                         batchProcessorsTracker,
                         batchTask,
                         key,
                         batchManagerId,
-                        profiler));
+                        profiler
+                )
+        );
+    }
+
+    private boolean waitForAvailableBatchProcessThread() {
+        log.trace("Wait available thread in BatchProcessor thread pool");
+        boolean isBatchProcessorThreadAvailable = false;
+        while (!isShutdown.get() && !isBatchProcessorThreadAvailable) {
+            try {
+                isBatchProcessorThreadAvailable =
+                        batchProcessorsTracker.tryAcquire(100, TimeUnit.MILLISECONDS);
+                if (isBatchProcessorThreadAvailable) {
+                    log.trace("isBatchProcessorThreadAvailable");
+                    batchProcessorsTracker.release();
+                }
+            } catch (InterruptedException exc) {
+                log.trace(THREAD_INTERRUPTED_MESSAGE, exc);
+                Thread.currentThread().interrupt();
+                return true;
             }
         }
-        log.trace("BatchProcessorManager thread stopped.");
+        return false;
+    }
+
+    @Nullable
+    private Map.Entry<KeyT, Queue<Operation<PayloadT>>> takeAnyQueueReadyForProcessing() {
+        return pendingTableOperations.entrySet().stream()
+                .peek(entry ->
+                        metricsProvider.createOperationsQueueSizeIndicatorIfNeeded(
```

Comment:
Seems like we are creating complex indicator tracking logic just to please inconvenient API of the profiler.  
Actually you are doing same logic of tracking indicator twice. 
1. you are tracking all queues in BatchingManager, carefully adding and removing indicators in profiler. 
2. then profiler does the same job buy tracking all indicators to build a report
Suggestion:  extend api of the Profiler instead. 
```
//exist: 
void attachIndicator(String name, IndicationProvider indicationProvider);
//add new one: 
void attachIndicators(IndicationsProvider indicationProvider);
void removeIndicaotrs(IndicationsProvider indicationProvider);
@FunctionalInterface
public interface IndicationsProvider {
    Map<Identity, Long> get() throws Exception;
}
Now we will have indicators that can provide dynamic names
```

Reply:
It looks like a significant aggregating profiler library revision... Can we leave it as is here for now with creating an issue in aggregating profiler repository for extend its functionality?